pingrid/impl.py

```diff
@@ -402,6 +412,69 @@ def produce_shape_tile(
     return im
 
 
+AQUAMARINE = BGRA(blue=212, green=255, red=127, alpha=255)
+BLUE = BGRA(blue=255, green=0, red=0, alpha=255)
+DARKORANGE = BGRA(blue=0, green=140, red=255, alpha=255)
+DARKRED = BGRA(blue=0, green=0, red=128, alpha=255)
+DEEPSKYBLUE = BGRA(blue=255, green=191, red=0, alpha=255)
+MOCCASIN = BGRA(blue=181, green=228, red=255, alpha=255)
+NAVY = BGRA(blue=128, green=0, red=0, alpha=255)
+PALEGREEN = BGRA(blue=152, green=251, red=152, alpha=255)
+RED = BGRA(blue=0, green=0, red=255, alpha=255)
+YELLOW = BGRA(blue=0, green=255, red=255, alpha=255)
+
+
+CORRELATIONCS = {
+    NAVY: [-1],
+    BLUE: [-0.8],
+    DEEPSKYBLUE: [-0.6],
+    AQUAMARINE: [-0.3],
+    PALEGREEN: [-0.1],
+    MOCCASIN: [-0.1, 0.1],
+    YELLOW: [0.1],
+    DARKORANGE: [0.4],
+    RED: [0.7],
+    DARKRED: [1],
+}
```

Comment:
A dictionary doesn't seem like the best representation. When you iterate through the items in the dictionary,  you get them in the order they were inserted, which is not necessarily the same as the order of the values. What happens if someone accidentally/unknowingly inserts values in the wrong order?

Let's define a ColorScale class, so we can put validation logic in the constructor and catch errors like that. The first argument could be a list of colors, and the second a list of thresholds, which is optional.

Secondarily, do we really need the special case where MOCCASIN gets two thresholds instead of one? Could we instead just repeat MOCCASIN twice in succession, with two different thresholds?

I don't know how familiar you are with defining classes. Let me know if you want more guidance.

Reply:
I will give it a shot. I could not repeat MOCCASSIN with the dictionary format.

Reply:
Oh, right. Another reason not to use a dictionary.

Reply:
On a side note: are we not reinventing the wheel here? Even if matplotlib is slow to make maps, we could still use it just for colors and colorscales, couldn't we?

Reply:
It's a reasonable question. Some factors to consider:
- There is a maintenance burden either way. If we write our own code, we have to debug it and extend it ourselves, but if we use someone else's we have to upgrade periodically and test the upgrades to make sure they don't break anything. Generally if it's simple it's better to write it ourselves, if it's complicated and there's a popular, mature alternative then it's better to use that.
- Matplotlib's colormap functionality is quite full-featured, whereas opencv's is more limited. If we want to translate matplotlib colormaps in their full generality to opencv then it seems like the translation code, which we own, will have to be quite complex. If we want to restrict ourselves to a limited subset of matplotlib's functionality in order to keep the complexity down, the best way to do that is to write our own colormap class that supports only the functionality we want to allow.
- All else being equal, I would prefer to leave options open rather than committing to using any particular library. If we tell people that the way to create a colormap in python maprooms is to import matplotlib, then we can't drop matplotlib later without breaking everyone's code. If we tell them to use a pingrid class, then even if that class uses matplotlib behind the scenes, we can drop matplotlib and reimplement the pingrid class later.

Sometimes we will expose a third-party API directly rather than hiding it from our users because it's so extensive and powerful that there's no reasonable way to hide it. This is the case with xarray, for example. It's not clear to me yet whether matplotlib is in that category, but until we have evidence that it is, I'd like to assume that it's not.

Reply:
I gave it a shot. Still meant to applying to the CORRELATION colorscale used in flex. fcst maproom. I kept the function I created previously as I barely know what I am doing.
Will clean up and apply to all colorscales once we are done with that one as case study